Offer pluggable serializers

roshan opened the pull request:
This allows using:

* Serializers that Permit backwards compatiblity
* Thrift serialization
* JSON serialization

Commit:
Offer pluggable serializers

This allows using:

* Serializers that Permit backwards compatiblity
* Thrift serialization
* JSON serialization

roshan commented on src/main/java/com/liveramp/daemon_lib/executors/JobletExecutors.java:
```diff
@@ -40,13 +41,17 @@
+      return get(notifier, tmpPath, maxProcesses, jobletFactoryClass, envVariables, successCallback, failureCallback, jobletRunner, JobletConfigStorage.DEFAULT_SERIALIZER, JobletConfigStorage.getDefaultDeserializer());
+    }
+
+    public static <T extends JobletConfig> ForkedJobletExecutor<T, JobletConfigMetadata, Integer> get(DaemonNotifier notifier, String tmpPath, int maxProcesses, Class<? extends JobletFactory<T>> jobletFactoryClass, Map<String, String> envVariables, JobletCallback<? super T> successCallback, JobletCallback<? super T> failureCallback, ProcessJobletRunner<Integer> jobletRunner, Function<? super T, b ...
```
Only 245 params more to go before we hit the JVM limit.

bpodgursky replied:
![image](https://user-images.githubusercontent.com/706280/29722462-c5691ef2-8975-11e7-9883-007ad03eead8.png)

pwestling replied:
If we run out of params, we can write the "full" sig in a comment and write a ruby script to parse the comment into a series of `Pair`s with the appropriate generic types ◉ ͜ʖ ◉

roshan commented on src/main/java/com/liveramp/daemon_lib/builders/BaseForkingDaemonBuilder.java:
```diff
@@ -46,11 +50,17 @@ public E addToEnvironmentVariables(String envVar, String value) {
   protected JobletExecutor<T> getExecutor() throws IllegalAccessException, IOException, InstantiationException {
     final String tmpPath = new File(workingDir, identifier).getPath();
-    return JobletExecutors.Forked.get(notifier, tmpPath, maxProcesses, jobletFactoryClass, envVariables, successCallback, failureCallback, jobletRunner);
+    return JobletExecutors.Forked.get(notifier, tmpPath, maxProcesses, jobletFactoryClass, envVariables, successCallback, failureCallback, jobletRunner, serializer, deserializer);
```
oh fuck, `deserializer` isn't initialised!

roshan replied:
This is fixed now.

pwestling reviewed (commented):
I guess my biggest question is if we imagine this being useful in a way that simply implementing a `JobletConfig`'s `readFields` and `writeFields` methods could not be? I guess there's no harm except for the increase in params and options so I'm not sure I feel strongly either way.

Part of me wants to say this is a large enough hole in the framework though that we should change `JobletConfig` to require providing a serializer/deserializer. Migrations are super painful right now and people aren't even aware of when they're breaking things.

roshan commented:
That's fair. We could just make `JobletConfig` `Externalizable` and require everyone to provide serialization/deserialization.

roshan commented:
I guess I could also just do that for my stuff and not require this change.

pwestling commented:
I guess my thought is - 
this is great if we want to change a lot of stuff to serialize with GSON, for example
this is not that great for changing everyone to thrift, because you still can't change the serializer without changing the joblet config

lancehc commented:
How is the `ForkedJobletRunner` going to know which deserializer to use?  That was the main problem I ran into when I tried to do this.  It seemed like you have to make a copy of the `ForkedJobletRunner` for each serialization scheme.  That doesn't work particularly well with something like a composite deserializer.  However, you could get away with having a composite deserializer in the daemon, and using a singular one in the joblet, since the joblet will never be reading stale configs.

Attaching the serialization method to the `JobletConfig` seems like it ought to solve that problem though.

Doing it with thrift configs would be tough.  It seems like you could make it work with a generic joblet config that wraps a thrift config, provided you had a way to get it the class name of the wrapped thrift config in the `ForkedJobletRunner`.  That doesn't seem very feasible though.

When I was poking at this last, I looked into gson serializing thrift structs, and it worked fine... provided they never rename the variables.

Commit:
Add deserialization support at ForkedJobletRunner

Commit:
Add Gson Serializer/Deserializer

roshan commented:
@lancehc I added deserialization to the `ForkedJobletRunner` in https://github.com/LiveRamp/daemon_lib/pull/9/commits/0fd990db62751c1942b192e02b81c5ee55b87701 by just passing along the classnames for the Deserializer. This may not work with a lambda, but honestly, I think that's all right. I added a Gson Serializer/Deserializer pair in https://github.com/LiveRamp/daemon_lib/pull/9/commits/9248cfbf2bf56a2fa4a29c47b9d15760d016ae58. Let me know what you think.

This particular impl has you pass in a deserializer to both your JobletRunner and to the Daemon. I have a separate branch where you only pass it in to the Daemon, but I'm not sure that's the right level to do things since only the forking runners actually need to know what deserializers to load. The non-forking versions have access to the `JobletConfigStorage` from the start.

lancehc commented on src/main/java/com/liveramp/daemon_lib/executors/forking/ForkedJobletRunner.java:
```diff
@@ -28,14 +33,30 @@ public static void main(String[] args) throws IOException, ClassNotFoundExceptio
-    JobletConfig config = JobletConfigStorage.production(configStorePath).loadConfig(id);
     DefaultJobletStatusManager jobletStatusManager = new DefaultJobletStatusManager(daemonWorkingDir);
+    final List<Function<byte[], JobletConfig>> deserializersWithDefault = Stream.concat(
+        Stream.of(JobletConfigStorage.getDefaultDeserializer()),
```
I think this would make more sense as the second argument (unless I'm misunderstanding the ordering with Stream.concat).  Otherwise, I imagine when you use a custom deserializer, it'll be erroring every time with the default one before trying the custom one.

roshan replied:
Mos def. Let me swappify.

lancehc reviewed (approved):
This is a really nice solution.  It seems like the migration ought to go very smoothly too.  :heart:

roshan commented:
Addressed and squashed.

@lancehc how do we usually go about merging these into `master`? Do I bump a the revision preceding the `SNAPSHOT` part?

pwestling commented:
We generally do not change the snapshot number - anyone external using a SNAPSHOT version should expect to be getting updates, and my assumption is that these changes don't break existing users interface-wise (correct me if I'm wrong there), so it shouldn't cause compilation failure chaos for anyone.

roshan commented:
All right. Here's how you add one of these things:

```
diff --git a/src/main/java/com/liveramp/daemon_lib/demo_daemon/DemoDaemon.java b/src/main/java/com/liveramp/daemon_lib/demo_daemon/DemoDaemon.java
index f753052..38a5bf5 100644
--- a/src/main/java/com/liveramp/daemon_lib/demo_daemon/DemoDaemon.java
+++ b/src/main/java/com/liveramp/daemon_lib/demo_daemon/DemoDaemon.java
@@ -1,5 +1,7 @@
 package com.liveramp.daemon_lib.demo_daemon;
 
+import java.util.Collections;
+
 import com.liveramp.daemon_lib.Daemon;
 import com.liveramp.daemon_lib.DaemonBuilders;
 import com.liveramp.daemon_lib.Joblet;
@@ -7,6 +9,8 @@ import com.liveramp.daemon_lib.JobletConfig;
 import com.liveramp.daemon_lib.JobletConfigProducer;
 import com.liveramp.daemon_lib.JobletFactory;
 import com.liveramp.daemon_lib.executors.forking.JarBasedProcessJobletRunner;
+import com.liveramp.daemon_lib.serialization.json.GsonDeserializer;
+import com.liveramp.daemon_lib.serialization.json.GsonSerializer;
 import com.liveramp.daemon_lib.utils.DaemonException;
 
 public class DemoDaemon {
@@ -62,8 +66,11 @@ public class DemoDaemon {
         new Producer(),
         JarBasedProcessJobletRunner
             .builder(".")
+            .addCustomDeserializers(Collections.singletonList(new GsonDeserializer<>()))
             .build()
     )
+        .setConfigStorageSerializer(new GsonSerializer<>(Config.class))
+        .addToConfigStorageDeserializers(new GsonDeserializer<>())
         .setMaxProcesses(4)
         .setConfigWaitSeconds(1)
         .setNextConfigWaitSeconds(1)
```
